src/compiler/debug.rs

```diff
@@ -20,11 +20,21 @@ impl DebugType {
         }
     }
 
-    pub fn assert_is_defined<TValue>(&self, value: Option<TValue>, message: Option<&str>) {
+    pub fn assert_is_defined<TValue>(&self, value: &Option<TValue>, message: Option<&str>) {
         if value.is_none() {
             self.fail(message);
         }
     }
+
+    pub fn check_defined<TValue>(&self, value: Option<TValue>, message: Option<&str>) -> TValue {
+        self.assert_is_defined(&value, message);
+        value.unwrap()
+    }
+
+    pub fn assert_never<TValue>(&self, value: TValue, message: Option<&str>) -> ! {
```

Comment:
What motivated the `!` return type here?

Reply:
In Typescript it's `never`, it'll always panic via `self.fail()` and so that helps it know about control flow at the point of calling `assert_never()`